### PR TITLE
fix(pnl): checking the side of trade, not taker or maker

### PR DESCRIPTION
**Pull request by jnlin**
The ask/bid amount should be calculated by the side of trade, not by taker or maker side

### CODE REVIEW

**Comment by c9s** [comment]
The IsBuy field is actually side. it's boolean field for optimization.

IsBuy = side == "BUY"

taker or maker depends on the boolean field IsMaker

**Comment by jnlin** [comment]
The 2 results are different. The definition of trade.isBuyer in Max
Exchange is "Side == bid" (Ref:
https://github.com/c9s/bbgo/blob/ed85465e71e7826623bb1ddee879a217282de68f/pkg/exchange/max/maxapi/trade.go#L39-L41
)

But the definition of trade.Side in Max Exchange is "Side == bid || Side ==
buy" (Ref:
https://github.com/c9s/bbgo/blob/ed85465e71e7826623bb1ddee879a217282de68f/pkg/exchange/max/convert.go#L37-L38
)

* t.isBuyer

```
[0000]  INFO loaded max server timestamp: 1612532106 offset=-1 exchange=max
[0000]  INFO SELECT * FROM trades WHERE exchange = :exchange AND symbol =
:symbol ORDER BY gid ASC LIMIT 7000
[0000]  INFO 6310 trades loaded
[0000]  INFO found checkpoints: [366]
[0000]  INFO stock: 0.000000
[0000]  INFO querying kline BTCUSDT 1m {Limit:100 StartTime:2021-02-05
19:55:06.567541744 +0800 CST m=-5999.700805831 EndTime:2021-02-05
21:35:06.567541744 +0800 CST m=+0.299194169} exchange=max
[0000]  INFO TRADES SINCE: 2020-11-12 16:31:39.513 +0000 UTC
[0000]  INFO NUMBER OF TRADES: 6310
[0000]  INFO AVERAGE COST: $ 24,261.98
[0000]  INFO TOTAL BUY VOLUME: 0.025966
[0000]  INFO TOTAL SELL VOLUME: 5.873472
[0000]  INFO STOCK: -5.847506
[0000]  INFO FEE (USD): 2.010891
[0000]  INFO CURRENT PRICE: $ 37,850.00
[0000]  INFO CURRENCY FEES:
[0000]  INFO  - MAX: 359.932906
[0000]  INFO  - BTC: 0.000034
[0000]  INFO  - USDT: 0.916735
[0000]  INFO PROFIT: $ 18,154.76
[0000]  INFO UNREALIZED PROFIT: $ 18,154.76
```

* t.Side == SideTypeBuy

```
[0000]  INFO loaded max server timestamp: 1612517136 offset=-1 exchange=max
[0000]  INFO SELECT * FROM trades WHERE exchange = :exchange AND symbol =
:symbol ORDER BY gid ASC LIMIT 7000
[0000]  INFO 6303 trades loaded
[0000]  INFO found checkpoints: [366]
[0000]  INFO stock: 0.000000
[0000]  INFO querying kline BTCUSDT 1m {Limit:100 StartTime:2021-02-05
15:45:37.142882886 +0800 CST m=-5999.685384032 EndTime:2021-02-05
17:25:37.142882886 +0800 CST m=+0.314615968} exchange=max
[0000]  INFO TRADES SINCE: 2020-11-12 16:31:39.513 +0000 UTC
[0000]  INFO NUMBER OF TRADES: 6303
[0000]  INFO AVERAGE COST: $ 27,164.03
[0000]  INFO TOTAL BUY VOLUME: 2.935320
[0000]  INFO TOTAL SELL VOLUME: 2.957118
[0000]  INFO STOCK: -0.021798
[0000]  INFO FEE (USD): 2.010891
[0000]  INFO CURRENT PRICE: $ 37,536.27
[0000]  INFO CURRENCY FEES:
[0000]  INFO  - MAX: 359.394043
[0000]  INFO  - BTC: 0.000034
[0000]  INFO  - USDT: 0.916735
[0000]  INFO PROFIT: $ 960.91
[0000]  INFO UNREALIZED PROFIT: $ 960.91
```

**Comment by elena-ann** [comment]
I got your point, the side calculation in max is tricky, I think the max
version has some problems. we might need to fix it

Jui-Nan Lin <notifications@github.com>於 2021年2月5日 週五，21:38寫道：

> The 2 results are different. The definition of trade.isBuyer in Max
> Exchange is "Side == bid" (Ref:
>
> https://github.com/c9s/bbgo/blob/ed85465e71e7826623bb1ddee879a217282de68f/pkg/exchange/max/maxapi/trade.go#L39-L41
> )
>
> But the definition of trade.Side in Max Exchange is "Side == bid || Side ==
> buy" (Ref:
>
> https://github.com/c9s/bbgo/blob/ed85465e71e7826623bb1ddee879a217282de68f/pkg/exchange/max/convert.go#L37-L38
> )
>
> * t.isBuyer
>
> ```
> [0000] INFO loaded max server timestamp: 1612532106 offset=-1 exchange=max
> [0000] INFO SELECT * FROM trades WHERE exchange = :exchange AND symbol =
> :symbol ORDER BY gid ASC LIMIT 7000
> [0000] INFO 6310 trades loaded
> [0000] INFO found checkpoints: [366]
> [0000] INFO stock: 0.000000
> [0000] INFO querying kline BTCUSDT 1m {Limit:100 StartTime:2021-02-05
> 19:55:06.567541744 +0800 CST m=-5999.700805831 EndTime:2021-02-05
> 21:35:06.567541744 +0800 CST m=+0.299194169} exchange=max
> [0000] INFO TRADES SINCE: 2020-11-12 16:31:39.513 +0000 UTC
> [0000] INFO NUMBER OF TRADES: 6310
> [0000] INFO AVERAGE COST: $ 24,261.98
> [0000] INFO TOTAL BUY VOLUME: 0.025966
> [0000] INFO TOTAL SELL VOLUME: 5.873472
> [0000] INFO STOCK: -5.847506
> [0000] INFO FEE (USD): 2.010891
> [0000] INFO CURRENT PRICE: $ 37,850.00
> [0000] INFO CURRENCY FEES:
> [0000] INFO - MAX: 359.932906
> [0000] INFO - BTC: 0.000034
> [0000] INFO - USDT: 0.916735
> [0000] INFO PROFIT: $ 18,154.76
> [0000] INFO UNREALIZED PROFIT: $ 18,154.76
> ```
>
> * t.Side == SideTypeBuy
>
> ```
> [0000] INFO loaded max server timestamp: 1612517136 offset=-1 exchange=max
> [0000] INFO SELECT * FROM trades WHERE exchange = :exchange AND symbol =
> :symbol ORDER BY gid ASC LIMIT 7000
> [0000] INFO 6303 trades loaded
> [0000] INFO found checkpoints: [366]
> [0000] INFO stock: 0.000000
> [0000] INFO querying kline BTCUSDT 1m {Limit:100 StartTime:2021-02-05
> 15:45:37.142882886 +0800 CST m=-5999.685384032 EndTime:2021-02-05
> 17:25:37.142882886 +0800 CST m=+0.314615968} exchange=max
> [0000] INFO TRADES SINCE: 2020-11-12 16:31:39.513 +0000 UTC
> [0000] INFO NUMBER OF TRADES: 6303
> [0000] INFO AVERAGE COST: $ 27,164.03
> [0000] INFO TOTAL BUY VOLUME: 2.935320
> [0000] INFO TOTAL SELL VOLUME: 2.957118
> [0000] INFO STOCK: -0.021798
> [0000] INFO FEE (USD): 2.010891
> [0000] INFO CURRENT PRICE: $ 37,536.27
> [0000] INFO CURRENCY FEES:
> [0000] INFO - MAX: 359.394043
> [0000] INFO - BTC: 0.000034
> [0000] INFO - USDT: 0.916735
> [0000] INFO PROFIT: $ 960.91
> [0000] INFO UNREALIZED PROFIT: $ 960.91
> ```
>

**Comment by c9s** [comment]
I think so too

**Comment by c9s** [comment]
in the max exchange implementation:
```
	return &types.Trade{
		ID:            int64(t.ID),
		OrderID:       t.OrderID,
		Symbol:        toGlobalSymbol(t.Market),
		Exchange:      "max",
		Price:         price,
		Quantity:      quantity,
		Side:          side,
		IsBuyer:       side == "bid",
		IsMaker:       t.Maker,
		Fee:           fee,
		FeeCurrency:   toGlobalCurrency(t.FeeCurrency),
		QuoteQuantity: quoteQuantity,
		Time:          mts,
	}, nil
```

**Comment by c9s** [comment]
the conversion looks correct: (max)

```
	return &types.Trade{
		ID:            int64(t.ID),
		OrderID:       t.OrderID,
		Price:         price,
		Symbol:        toGlobalSymbol(t.Market),
		Exchange:      "max",
		Quantity:      quantity,
		Side:          side,
		IsBuyer:       t.IsBuyer(),
		IsMaker:       t.IsMaker(),
		Fee:           fee,
		FeeCurrency:   toGlobalCurrency(t.FeeCurrency),
		QuoteQuantity: quoteQuantity,
		Time:          mts,
	}, nil
```

side:

```
var side = toGlobalSideType(t.Side)
```

and `toGlobalSideType`:

```
func toGlobalSideType(v string) types.SideType {
	switch strings.ToLower(v) {
	case "bid", "buy":
		return types.SideTypeBuy

	case "ask", "sell":
		return types.SideTypeSell

	case "self-trade":
		return types.SideTypeSelf

	}

	return types.SideType(v)
}
```

and the IsBuyer, IsMaker method:

```
func (t Trade) IsBuyer() bool {
	return t.Side == "bid"
}

func (t Trade) IsMaker() bool {
	return t.Info.Maker == t.Side
}
```

**Comment by c9s** [comment]
maybe we should change IsBuyer to the following code:

```
func (t Trade) IsBuyer() bool {
	return t.Side == "bid" || t.Side == "buy"
}
```

the reason is that MAX API sometimes replies "buy" and sometimes it replies "bid" ....... orz

File: pkg/accounting/pnl/avg_cost.go [hunk 1]
```diff
@@ -36,7 +36,7 @@ func (c *AverageCostCalculator) Calculate(symbol string, trades []types.Trade, c
 
 	for _, trade := range trades {
 		if trade.Symbol == symbol {
-			if trade.IsBuyer {
+			if trade.Side == types.SideTypeBuy {
```
Comment (c9s) [inline pkg/accounting/pnl/avg_cost.go:39]: For MAX, it's possible to get types.SideTypeSelf, and it's not considered here. 

The code was originally for binance.

**Comment by c9s** [comment]
btw, for the position calculation, you can take a look at bbgo.Position, see if we can reuse.

**Comment by c9s** [comment]
fyi: you can join our telegram channel https://t.me/bbgocrypto

File: pkg/accounting/pnl/avg_cost.go (hunk 1 above)
Comment (jnlin) [reply pkg/accounting/pnl/avg_cost.go]: Should be fixed

**Comment by jnlin** [comment]
> maybe we should change IsBuyer to the following code:
> 
> ```
> func (t Trade) IsBuyer() bool {
> 	return t.Side == "bid" || t.Side == "buy"
> }
> ```
> 
> the reason is that MAX API sometimes replies "buy" and sometimes it replies "bid" ....... orz

Fixed with this patch. Should re-sync all trade records from max exchange

**Comment by c9s** [comment]
yes 👍 thanks! 🤩

Jui-Nan Lin <notifications@github.com>於 2021年2月6日 週六，18:15寫道：

**Review by c9s — APPROVED** [review]

**Pull request merged**